xwiki-platform-core/xwiki-platform-realtime/xwiki-platform-realtime-wysiwyg/xwiki-platform-realtime-wysiwyg-test/xwiki-platform-realtime-wysiwyg-test-docker/src/test/it/org/xwiki/realtime/wysiwyg/test/ui/RealtimeWYSIWYGEditorIT.java

```diff
@@ -1336,6 +1337,205 @@ void editPageTranslations(TestUtils setup, TestReference testReference, MultiUse
+        
+        // Check that the second user did not re-join the realtime editing session.
+        assertFalse(secondEditPage.isRealtimeEditing());
+        assertTrue(secondEditorToolbar.canToggleSourceMode());
+        assertEquals("one\ntwo\nthree\nfour\nfive", secondTextArea.getText());
+        
+        // Join the realtime session again and wait to be in sync.
+        secondEditPage.joinRealtimeEditing();
+        secondTextArea.waitUntilContentContains("six");
+        assertEquals("one\ntwo\nthree\nfour\nsix", secondTextArea.getText());
+        
+        // Check that we can still switch to source mode.
+        assertTrue(secondEditorToolbar.canToggleSourceMode());
+        
+        // Make the editor dirty by editing the content without saving.
+        secondTextArea.sendKeys(Keys.ARROW_DOWN, Keys.END, Keys.ENTER, "seven");
+        
+        // Switch to source mode and back to wysiwyg edit mode.
+        secondEditorToolbar.toggleSourceMode();
+        assertFalse(secondEditPage.isRealtimeEditing());
+        assertEquals("one\n\ntwo\n\nthree\n\nfour\n\nsix\n\nseven", secondEditor.getSourceTextArea().getAttribute("value"));
+        secondEditorToolbar.toggleSourceMode();
+        
+        // Check that the second user did not re-join the realtime editing session.
+        assertFalse(secondEditPage.isRealtimeEditing());
+        
+        // We keep the second user out of the realtime editing session now
+        // and we do more tests with the first user.
+        
+        //
+        // First Tab
+        //
+
+        // Switch to the first tab.
+        setup.getDriver().switchTo().window(multiUserSetup.getFirstTabHandle());
+
+        // Check that the source button is available.
+        assertTrue(firstEditorToolbar.canToggleSourceMode());
+        
+
+        
+        // Saving might merge silently. Here is some bulletproofing cleanup.
+        firstEditPage.clickSaveAndContinue();
+        firstTextArea.clear();
+        
+        // Make the editor not dirty by saving.
+        firstEditPage.clickSaveAndContinue();        
+        
+        // Switch to source mode and back to wysiwyg.
+        firstEditorToolbar.toggleSourceMode();
+        assertFalse(firstEditPage.isRealtimeEditing());
+        firstEditorToolbar.toggleSourceMode();
+        assertTrue(firstEditPage.isRealtimeEditing());
+        
+        // Bulletproofing: Save to make sure the editor is not dirty.
+        firstEditPage.clickSaveAndContinue();
+        
+        // Switch to source mode, make a change, and switch back to wysiwyg.
+        firstEditorToolbar.toggleSourceMode();
+        assertFalse(firstEditPage.isRealtimeEditing());
+        firstEditor.getSourceTextArea().sendKeys(Keys.ENTER, Keys.ENTER, "eight");
+        firstEditorToolbar.toggleSourceMode();
+        
+        firstTextArea = firstEditor.getRichTextArea();
+        assertTrue(firstEditPage.isRealtimeEditing());
```

Comment:
Maybe add a comment explaining that we joined the realtime session because we're editing alone.